### PR TITLE
Generate c code

### PR DESCRIPTION
**Pull request by Kgeorgouli**
I have amended the current PR by removing the last commit with the testing files.

### CODE REVIEW

**Comment by JaeseungYeom** [comment]
It could have been better if there were two separate PRs, one commit per PR. Fortunately, commits are clearly distinctive. So, it is ok.
I think we do not need the model output files since those can be easily reproduced by running the code.

**Review by JaeseungYeom — CHANGES_REQUESTED** [review]
I sent you a file with all of my suggested updates applied.

File: src/utils/generate_C_code.cpp [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+                   model_reactions_s);
+
+  std::vector<std::string>::const_iterator it, nit;
+  //remove duplicates
+  for  (it = dependencies.cbegin(); it < dependencies.cend(); it++) {
+    ndit = dependencies_set.find (*it);
+    if (ndit == dependencies_set.cend()) {
+      dependencies_set.insert(*it);
+      nit = dependencies_no_dupl.begin();
+      dependencies_no_dupl.insert(nit, *it);
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: replace nit with dependencies_no_dupl.begin()

File: src/utils/generate_C_code.cpp [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+
+#if defined(WCS_HAS_CONFIG)
+#include "wcs_config.hpp"
+#else
+#error "no config"
+#endif
+
+#include <unordered_map>
+#include "sbml_utils.hpp"
+#include "cstdio"
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: replace "cstdio" with \<cstdio\>

File: src/utils/generate_C_code.cpp [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+  std::unordered_set<std::string>::const_iterator ndit;
+
+  get_dependencies(formula,
+                   dependencies,
+                   good,
+                   constant_init_assig,
+                   variables_assig_rul,
+                   model_reactions_s);
+
+  std::vector<std::string>::const_iterator it, nit;
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: we do not need `nit` with the change applied to line 131.

**Review by JaeseungYeom — APPROVED** [review]
Just a couple of more changes.

File: src/utils/generate_C_code.cpp [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+    unsigned int num_arg = function.getNumArguments();
+
+    for (unsigned int ia = 0u; ia < num_arg; ia++) {
+      std::cout << "Real " << function.getArgument(ia)->getName();
+      if (ia!=num_arg-1) {
+        std::cout <<", ";
+      }
+    }
+
+    std::cout <<") {\n";
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: space after `<<`

File: src/utils/generate_C_code.cpp [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+    std::cout << "  return " << SBML_formulaToString(function.getBody()) <<";";
+    std::cout <<"\n}\n\n";
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: space after `<<`

File: src/utils/generate_C_code.cpp [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+    const LIBSBML_CPP_NAMESPACE::FunctionDefinition& function
+      = *(function_definition_list->get(ic));
+
+    std::cout << "static inline Real " << function.getIdAttribute() <<"(";
+    unsigned int num_arg = function.getNumArguments();
+
+    for (unsigned int ia = 0u; ia < num_arg; ia++) {
+      std::cout << "Real " << function.getArgument(ia)->getName();
+      if (ia!=num_arg-1) {
+        std::cout <<", ";
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: space after `<<`

File: src/utils/generate_C_code.cpp [hunk 7]
```diff
@@ -0,0 +1,715 @@
...
+  for (unsigned int ic = 0u; ic < functionsSize; ic++) {
+    const LIBSBML_CPP_NAMESPACE::FunctionDefinition& function
+      = *(function_definition_list->get(ic));
+
+    std::cout << "static inline Real " << function.getIdAttribute() <<"(";
+    unsigned int num_arg = function.getNumArguments();
+
+    for (unsigned int ia = 0u; ia < num_arg; ia++) {
+      std::cout << "Real " << function.getArgument(ia)->getName();
+      if (ia!=num_arg-1) {
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: space around `!=`

File: src/utils/generate_C_code.cpp [hunk 8]
```diff
@@ -0,0 +1,715 @@
...
+    }
+  }
+
+
+  /// Print out Species initializations
+  for (unsigned int ic = 0u; ic < speciesSize; ic++) {
+    const LIBSBML_CPP_NAMESPACE::Species& species = *(species_list->get(ic));
+
+    if (species.getInitialAmount() == 0) {
+      std::cout << "static constexpr Real _init_" << species.getIdAttribute() << " = 0;\n";
```
Comment (JaeseungYeom) [inline src/utils/generate_C_code.cpp]: break this long line into two

**Review by JaeseungYeom — APPROVED** [review]
and some more

**Comment by JaeseungYeom** [comment]
Maybe the file name should be `generate_cxx_code.cpp` because it generate c++ code.

**Comment by JaeseungYeom** [comment]
To fix the conflict in CMakeList.txt, change WCS_UTILS_TARGETS to WCS_EXEC_TARGETS

**Comment by Kgeorgouli** [comment]
I have done all the changes as you requested.

**Comment by JaeseungYeom** [comment]
Resolve the conflict and then merge it. Thanks a lot!
It still shows up as unresolved because it does not know what to do with the new lines added in your branch in CMakeList.txt
You can keep those lines. You just need to remove the three marker lines added by git.
```
<<<< generate_c_code
=================
>>>> devel
```
I fixed it.

**Comment by Kgeorgouli** [comment]
I made the rebase and fixed the conflict as you suggested.

File: CMakeLists.txt [hunk 9]
```diff
@@ -453,6 +461,18 @@ list(APPEND WCS_UNIT_TEST_TARGETS t_state_rngen-bin)
...
+  TARGETS  reaction-bin ssa-bin
+  EXPORT   WCSTargets
+  RUNTIME  DESTINATION ${CMAKE_INSTALL_BINDIR}
+  ARCHIVE  DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  LIBRARY  DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+  )
+
+install(
+>>>>>>> Generate C code from a SBML file
```
Comment (JaeseungYeom) [inline CMakeLists.txt]: remove these lines.

Comment (JaeseungYeom) [reply CMakeLists.txt]: > I made the rebase and fixed the conflict as you suggested.

That is great. Can you do once more by removing the lines shown above?

**Comment by Kgeorgouli** [comment]
Done!

**Comment by JaeseungYeom** [comment]
You can merge it now. If you still want to perfect it further, there are some invisible tabs '\t', which you can remove or replace with space.

**Comment by Kgeorgouli** [comment]
I removed  also the invisible tabs but when I try to run the cmake, I take the following error "CMake Error: install(EXPORT "WCSTargets" ...) includes target "reaction-bin" more than once in the export set.
CMake Error in CMakeLists.txt:
  given target "reaction-bin" more than once."

File: CMakeLists.txt (hunk 9 above)
Comment (JaeseungYeom) [reply CMakeLists.txt]: In the new file, it is around lines 463-471. Those need to be removed.

**Comment by JaeseungYeom** [comment]
> I removed also the invisible tabs but when I try to run the cmake, I take the following error "CMake Error: install(EXPORT "WCSTargets" ...) includes target "reaction-bin" more than once in the export set.
> CMake Error in CMakeLists.txt:
> given target "reaction-bin" more than once."

The install block for `reaction-bin ssa-bin` need to be removed.
It is because reaction-bin ssa-bin as well as all other *-bin are now stored in WCS_EXEC_TARGETS, and there is a install block for this. So, it will install all *-bin.

**Comment by Kgeorgouli** [comment]
I have removed the install block for  reaction-bin ssa-bin as you requested and everything is okay now.

**Comment by JaeseungYeom** [comment]
> I have removed the install block for reaction-bin ssa-bin as you requested and everything is okay now.

Excellent. Hit the green button!

**Comment by Kgeorgouli** [comment]
> > I have removed the install block for reaction-bin ssa-bin as you requested and everything is okay now.
> 
> Excellent. Hit the green button!

Thank you very much for your help!

**Pull request merged**